Type check with pytype

AdamGleave opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

  - Add type checking with `pytype` to the Travis CI builds; update Dockerfile accordingly.
  - Add `py.typed` file to declare PEP 561 compliance.
  - Fix a few minor bugs caught by `pytype`: mostly wrong variable names.
  - Suppress some false-positives or adapt the code so that it can be type checked.

Note what I have *not* done is added any type annotations (except where it was needed to make code type check). I don't have the bandwidth to do this right now, although it'd be great if we added annotations to some of the core classes.

<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- You can use the syntax `closes #100` if this solves the issue #100 -->

Closes #561.
- [X] I have raised an issue to propose this change ([required](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I've read the [CONTRIBUTION](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) guide (**required**)
- [X] I have updated the [changelog](https://github.com/hill-a/stable-baselines/blob/master/docs/misc/changelog.rst) accordingly (**required**).
- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

<!--- This Template is an edited version of the one from https://github.com/evilsocket/pwnagotchi/ -->

AdamGleave commented:
@araffin can you rebuild the Docker image to include `pytype`? (Or give me access to DockerHub so I can push a new image.)

araffin commented:
@AdamGleave sorry for that. I'll try to do that (not sure if i can with the free version)
in the mean time, you can change the travis config to point to your image

araffin commented:
@AdamGleave what is your docker username?

AdamGleave commented:
> @AdamGleave what is your docker username?

I'm [https://hub.docker.com/u/adamgleave](adamgleave). Thanks!

AdamGleave commented:
Travis is now passing. I'm now done with this PR (except any review comments).

One thing we shouldn't forgot is to change `.travis.yml` to point to `araffin/stable-baselines-cpu` since I changed it to point to my account temporarily.

Codacy has some false positives for unused imports that are used in type hints in comments. I think we'll just have to ignore these: it'd probably understand if I used variable type annotations, but that's only supported in 3.6+.

araffin commented:
>but that's only supported in 3.6+.

Maybe a good time to drop 3.5 support? This would allow to use f-strings and faster dict implementation that keeps the insertion order. I have to check when is the official end of support of 3.5...

> I think we'll just have to ignore these

yes, otherwise, it is not a big deal to ignore those (you have now the rights to do that normally ;) )

Miffyli commented:
Regarding Python 3.5: Many older linux distros come with Python 3.5 by default and it is difficult-ish to upgrade the pre-packaged one to newer one, so that may be bit of a headache for some (E.g. They need the older distribution/packages to run some specific software). Then again, they could just install minoconda and go with that, so I would vote for moving to Python versions +3.6.

AdamGleave commented:
> Maybe a good time to drop 3.5 support? This would allow to use f-strings and faster dict implementation that keeps the insertion order. I have to check when is the official end of support of 3.5...
[End of life](https://devguide.python.org/#branchstatus) for 3.5 is 2020-09-13, although I'd be OK dropping support earlier.

araffin commented on stable_baselines/common/cmd_util.py:
```diff
@@ -5,11 +5,11 @@
 
 import gym
-from gym.wrappers import FlattenDictWrapper
+from gym.wrappers import FilterObservation, FlattenObservation
```
maybe a try/except for old gym version?

AdamGleave replied:
added

araffin commented on docs/misc/changelog.rst:
```diff
@@ -19,6 +19,8 @@ New Features:
 - `VecNormalize` now supports being pickled and unpickled.
 - Add parameter `exploration_initial_eps` to DQN. (@jdossgollin)
+- Add type checking and PEP 561 compliance.
+  Note: most functions are still not annotated, this will be a gradual process.
```
Missing: the fix for newer gym version

AdamGleave replied:
added

araffin commented on stable_baselines/deepq/dqn.py:
```diff
@@ -243,13 +244,16 @@ def learn(self, total_timesteps, callback=None, log_interval=100, tb_log_name="D
                     # Minimize the error in Bellman's equation on a batch sampled from replay buffer.
+                    # pytype:disable=bad-unpacking
                     if self.prioritized_replay:
+                        assert self.beta_schedule is not None
```
error message missing

AdamGleave replied:
added

araffin commented on stable_baselines/deepq/experiments/run_atari.py:
```diff
@@ -18,8 +18,6 @@ def main():
     parser.add_argument('--dueling', type=int, default=1)
     parser.add_argument('--prioritized-replay-alpha', type=float, default=0.6)
     parser.add_argument('--num-timesteps', type=int, default=int(10e6))
-    parser.add_argument('--checkpoint-freq', type=int, default=10000)
```
nice catch, at some we should get rid of those files anyway now that we have the zoo

AdamGleave commented on .travis.yml:
```diff
@@ -4,7 +4,7 @@ python:
 env:
   global:
-    - DOCKER_IMAGE=araffin/stable-baselines-cpu:v2.7.1
+    - DOCKER_IMAGE=adamgleave/stable-baselines-cpu:v2.9.0
```
I've made a `stablebaselines` org and added you as an owner.

araffin replied:
good initiative

AdamGleave commented on .gitignore:
```diff
@@ -19,13 +20,11 @@ keys/
 # Virtualenv
 /env
-
+/venv
```
It's duplicated (also present on line 9)

AdamGleave commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+    . $VENV/bin/activate && \
+    cd $CODE_DIR && \
+    pip install --upgrade pip && \
+    if [[ $USE_GPU == "True" ]]; then \
```
thanks

AdamGleave commented on scripts/build_docker.sh:
```diff
@@ -0,0 +1,16 @@
+CPU_PARENT=ubuntu:16.04
+GPU_PARENT=nvidia/cuda:9.0-cudnn7-runtime-ubuntu16.04
+
+TAG=araffin/stable-baselines
```
changed to stable baselines

AdamGleave commented on scripts/run_docker_cpu.sh:
```diff
@@ -6,7 +6,6 @@ cmd_line="$@"
-
 docker run -it --rm --network host --ipc=host \
- --mount src=$(pwd),target=/root/code/stable-baselines,type=bind araffin/stable-baselines-cpu:v2.7.1 \
+ --mount src=$(pwd),target=/root/code/stable-baselines,type=bind araffin/stable-baselines-cpu:v2.9.0 \
```
changed to stable baselines

AdamGleave commented on stable_baselines/common/bit_flipping_env.py:
```diff
@@ -98,7 +98,10 @@ def step(self, action):
+    def compute_reward(self,
+                       achieved_goal: np.ndarray,
+                       desired_goal: np.ndarray,
+                       _info) -> int:
```
OK, I've updated the constant values to be floats not ints for consistency.

AdamGleave commented:
@araffin thanks for the review, I've addressed the comments.

AdamGleave commented:
While we're making some changes to the CI, wondering if we should disable Travis branch builds, and just build PRs? Right now our build queue in Travis is often quite long.

We'd need to change `scripts/run_tests_travis.sh` since right now the Codacy upload only happens on branch builds. But this is just becuase `CODACY_PROJECT_TOKEN` may not be available on forks. We could instead directly check for the presence of that environment variable.

araffin commented:
I agree on both ;)

AdamGleave commented:
> Almost there, still some edit needed in the doc: https://stable-baselines.readthedocs.io/en/master/guide/install.html#build-the-docker-images

Sorry for missing, updated.

> + maybe a line about running type check?

I've added this in `CONTRIBUTING.md`, couldn't see a better place for it.

> EDIT: I ignored the issues in codacy related to typing

Thanks :)

> EDIT: I'm afraid only @hill-a has the right to change that (for the build)

Travis seems to have let me disable branch builds so I've done that for now (and updated `scripts/run_tests_travis.sh` accordingly).

AdamGleave commented:
We seem to have both a `travis-ci.com` and `travis-ci.org` instance! The `travis-ci.com` is the one producing "Travis CI - Pull Request " and `travis-ci.org` is the one with "continuous-integration/travis-ci/pr" and "continuous-integration/travis-ci/push".

`travis-ci.org` is deprecated so it's probably best if we just get rid of that one. I've disabled builds on it but can't seem to delete it. @hill-a or someone else with admin access to this repo on GitHub should follow [this](https://help.github.com/en/github/administering-a-repository/enabling-required-status-checks) to remove the status checks beginning with "continuous-integration/"

araffin commented:
Unfortunately, i think only @hill-a  has the rights to do that. Will ping him again if he does not see this message ;)
